TIP-2332: New databaseUIViewOnly project flag to dx-toolkit

rohvani opened the pull request:
This PR adds a new `--database-ui-view-only` project flag for `update` (which can be displayed via `describe`), as well as support for the flag in our `dxpy` bindings (for `new` and `update`)

The CLI parses the `database-ui-view-only` flag into `databaseUIViewOnly`, which is ultimately passed to the apiserver as an argument.

rohvani commented on src/python/dxpy/bindings/dxproject.py:
```diff
@@ -282,8 +282,9 @@ class DXProject(DXContainer):
-            restricted=None, download_restricted=None, contains_phi=None, tags=None,
-            properties=None, bill_to=None, **kwargs):
+            restricted=None, download_restricted=None, contains_phi=None, 
+            tags=None, properties=None, bill_to=None, database_ui_view_only=None,
```
I added the new `database_ui_view_only` flag here so that it can be set with our `dxpy` bindings.  

Something I noticed is that while `contains_phi` can be passed here, we don't seem to be doing this for `update` ([link](https://github.com/dnanexus/dx-toolkit/blob/95bfc3ed51383094eb597394ed65f24ea4fd6646/src/python/dxpy/bindings/dxproject.py#L346-L348)).

Is that intentional?  Should I include the `database_ui_view_only` flag in the `update` function?

durae replied:
Yes, please add `database_ui_view_only` also to update, similar to `restricted` and `download_restrcited`. I think omitting `contains_phi` was intentional because once project is set to contain PHI, it can never be updated to not contain them.

durae commented on src/python/dxpy/scripts/dx.py:
```diff
@@ -48,7 +48,7 @@
                            set_env_from_args, extra_args, process_extra_args, DXParserError, exec_input_args,
                            instance_type_arg, process_instance_type_arg, process_instance_count_arg, get_update_project_args,
-                           property_args, tag_args, contains_phi, process_phi_param)
+                           property_args, tag_args, contains_phi, database_ui_view_only, process_phi_param)
```
I think you don't need to import it here as it is not used directly in the file only as `args.database_ui_view_only`

durae commented on src/python/dxpy/scripts/dx.py:
```diff
@@ -4759,6 +4761,8 @@ def register_parser(parser, subparsers_action=None, categories=('other', ), add_
 parser_update_project.add_argument('--containsPHI', choices=["true"],
                                    help="Flag to tell if project contains PHI")
+parser_update_project.add_argument('--database-ui-view-only', choices=["true", "false"],
+                                   help="Whether the viewers on the project can access the database details directly")
```
details -> data
because metadata are still accessible not only from UI

durae commented on src/python/dxpy/scripts/dx.py:
```diff
@@ -5051,6 +5055,8 @@ def register_parser(parser, subparsers_action=None, categories=('other', ), add_
 parser_new_project.add_argument('--bill-to', help='ID of the user or org to which the project will be billed. The default value is the billTo of the requesting user.')
 parser_new_project.add_argument('--phi', help='Add PHI protection to project', default=False,
                                 action='store_true')
+parser_new_project.add_argument('--database-ui-view-only', help='Restrict viewers on the project from accessing the database details directly', default=False,
```
I don't think we need to add a support for `databaseUIViewOnly` flag also to project new. I would treat it in a same way as `restricted` and `downloadRestricted` flags as these flags can be updated. `containsPHI` flag can't be changed once it is set on a project because it's tight to PHI access restrictions so therefore it makes sense to have it here.

durae reviewed (commented):
Some comments and answers below, otherwise LGTM.

arunbhat commented:
@durae please have someone from the devex team to review too.  @eugenemi fyi..

rohvani commented:
I've gone ahead and pushed a commit making the changes you requested @durae.  Here is a list of the changes:

- Removed database-ui-view-only flag from new project command
- Added database_ui_view_only flag to dxpy update binding
- Updated database-ui-view-only help text
- Removed unused import

durae reviewed (commented):
LGTM. Were you able to try it with local build of dx-toolkit?

durae commented:
@kpjensen could you please review the PR? We want to support for a new project flag to update and describe calls. Thanks!

rohvani commented:
> LGTM. Were you able to try it with local build of dx-toolkit?

I was, here are the results of `dx describe` and `dx update project`:

```bash
root@8e5056858fb9:/dx# dx describe project-G0p36Y80pQv4jp4X1bqbgqGZ
Result 1:
ID                  project-G0p36Y80pQv4jp4X1bqbgqGZ
Class               project
Name                my test project
Summary             
Billed to           org-tip_dev
Access level        ADMINISTER
Region              aws:us-east-1
Protected           false
Restricted          false
Contains PHI        false
Database UI View Onlyfalse
Created             Mon Feb 22 22:42:21 2021
Created by          ruvanni1
Last modified       Wed May 26 00:17:22 2021
Data usage          2.70 GB
Sponsored data      0.00 GB
Storage cost        $0.101/month
Sponsored egress    0.00 GB used of 0.00 GB total
At spending limit?  false
Tags                -
Properties          -
currency            {"symbolPosition": "left", "groupingSymbol": ",", "decimalSymbol": ".",
                    "symbol": "$", "code": "USD", "dxCode": 0}
downloadRestricted  false
remoteDataUsage     0
defaultInstanceType "mem2_hdd2_x2"
archivedDataUsage   0
provider            {}
cloudAccount        "cloudaccount-dnanexus"

root@8e5056858fb9:/dx# dx update project project-G0p36Y80pQv4jp4X1bqbgqGZ --database-ui-view-only true
{"id": "project-G0p36Y80pQv4jp4X1bqbgqGZ"}

root@8e5056858fb9:/dx# dx describe project-G0p36Y80pQv4jp4X1bqbgqGZ
Result 1:
ID                  project-G0p36Y80pQv4jp4X1bqbgqGZ
Class               project
Name                my test project
Summary             
Billed to           org-tip_dev
Access level        ADMINISTER
Region              aws:us-east-1
Protected           false
Restricted          false
Contains PHI        false
Database UI View Onlytrue
Created             Mon Feb 22 22:42:21 2021
Created by          ruvanni1
Last modified       Wed May 26 00:17:50 2021
Data usage          2.70 GB
Sponsored data      0.00 GB
Storage cost        $0.101/month
Sponsored egress    0.00 GB used of 0.00 GB total
At spending limit?  false
Tags                -
Properties          -
currency            {"symbolPosition": "left", "groupingSymbol": ",", "decimalSymbol": ".",
                    "symbol": "$", "code": "USD", "dxCode": 0}
downloadRestricted  false
remoteDataUsage     0
defaultInstanceType "mem2_hdd2_x2"
archivedDataUsage   0
provider            {}
cloudAccount        "cloudaccount-dnanexus"

root@8e5056858fb9:/dx# 
```

arunbhat commented:
@kpjensen is this ok to merge?

@rohvani the Ux on the flag seems to be a bit off. Does it need a tab?

![image](https://user-images.githubusercontent.com/4691119/120526458-1335d700-c38e-11eb-86a6-d15b33edc914.png)

rohvani commented:
> @kpjensen is this ok to merge?
> 
> @rohvani the Ux on the flag seems to be a bit off. Does it need a tab?
> 
> ![image](https://user-images.githubusercontent.com/4691119/120526458-1335d700-c38e-11eb-86a6-d15b33edc914.png)

I also noticed the misalignment here, I will look into fixing this today.

rohvani commented:
I've gone ahead and updated the field name width so that the new database UI view flag doesn't overflow in the `dx describe` output:

 
![image](https://user-images.githubusercontent.com/3782201/120539735-be4d8d00-c39c-11eb-946b-ea9effa1a47f.png)

kpjensen commented on src/python/dxpy/bindings/dxproject.py:
```diff
@@ -385,6 +393,8 @@ def update(self, name=None, summary=None, description=None, protected=None,
         if unset_allowed_executables is not None:
             update_hash["allowedExecutables"] = None
+        if database_ui_view_only is not None:
+            input_hash["databaseUIViewOnly"] = database_ui_view_only
```
I think this should be `update_hash` here.

kpjensen commented on src/python/dxpy/utils/describe.py:
```diff
@@ -444,6 +444,8 @@ def print_project_desc(desc, verbose=False):
         print_json_field("Restricted", desc["restricted"])
     if 'containsPHI' in desc:
         print_json_field('Contains PHI', desc['containsPHI'])
+    if 'databaseUIViewOnly' in desc:
```
If this flag is not going to be publicized, do we want to print it out even when its set to false?

durae replied:
I agree, let's print it only when the flag is set to true. Thanks for the suggestion!

kpjensen commented on src/python/dxpy/cli/parsers.py:
```diff
@@ -281,6 +281,10 @@ def __call__(self, parser, namespace, values, option_string=None):
                           help='If set to true, only projects that contain PHI data will be retrieved. ' +
                           'If set to false, only projects that do not contain PHI data will be retrieved.')
 
+database_ui_view_only = argparse.ArgumentParser(add_help=False)
```
I don't think this needs flag needs its own parser. You can add `'--database-ui-view-only'` directly to `parser_new_project` like `parser_update_project` in `dx.py`

rohvani commented:
I've gone ahead and made the requested changes @kpjensen, please let me know what you think.  Thank you!

kpjensen reviewed (approved):
LGTM
